### PR TITLE
Windows: Skip failing cuDNN tests

### LINKED ISSUE #3872: CUDNN Tests fails in windows
We have observed that in windows with CUDNN 7.6 the following tests fail:

```
tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_93_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_94_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_95_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_381_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_382_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_383_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_669_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_670_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_671_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_957_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_958_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_959_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
```
`ConvolutionBackwardData`
The common parameters that fail for all tests are:
ndim=3, groups=2, dtype=float16, deterministic=True, dilate=2 and max_workspace_size=4194304

The GPU is a Tesla P4

### PR DESCRIPTION
**Pull request by leofang**
Close #3872.

Another guess work. My theory is the workspace size is too stringent for the failing tests:
```
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_93_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_94_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_95_{auto_tune=True, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_381_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_382_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_383_{auto_tune=True, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_669_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_670_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_671_{auto_tune=False, bias=True, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_957_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='always'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_958_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='auto'}::test_call
FAILED tests/cupy_tests/test_cudnn.py::TestConvolutionBackwardData_param_959_{auto_tune=False, bias=False, deterministic=True, dilate=2, dtype=float16, groups=2, max_workspace_size=4194304, ndim=3, tensor_core='never'}::test_call
```
It doesn't seem to be a good idea to me to hardcode the workspace size in order to choose an algorithm from `cudnnGetConvolutionBackwardDataAlgorithm_v7`:
https://github.com/cupy/cupy/blob/51ac662bc02a9291bead594a29d81efb872c4b1b/cupy/cudnn.pyx#L1598-L1599
as the performance result depends on the available workspace size used by the algorithm, among other things.
But admittedly I know nothing about cuDNN so there could be other reasons doing so...

### CODE REVIEW

**Comment by leofang** [comment]
Well, no luck 😂 Increasing the max workspace size by a factor of 2 does not help:
https://storage.googleapis.com/chainer-artifacts-pfn-public-ci/cupy-ci/70025/cupy_test_log.txt
Now I wonder why setting workspace size to 0 can pass the tests....

**Comment by leofang** [comment]
OK, so with 9927631 I was able to tell that some algorithms are found (hence the list is not empty). But it still can't explain why tests passed only with zero `max_workspace_size`.

**Comment by leofang** [comment]
> TODO: Follow #3790 and #4689 to disable the failing tests.

Done in 1128937.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
> But it still can't explain why tests passed only with zero `max_workspace_size`.

Ahh I overlooked this:
https://github.com/cupy/cupy/blob/2e95ed7e994b1db230501a69e8a323666164fb80/tests/cupy_tests/test_cudnn.py#L362-L364
so `max_workspace_size=0` is not tested...OK all is making sense now: at least one algorithm is found, but either it's non-deterministic, or it requires a large workspace.

**Comment by leofang** [comment]
Why isn't the tests skipped? I am clueless...

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Finally fixed...In the earlier commits CuDNNerror was raised, not RuntimeError, so we couldn't skip. It's odd that unittest does not complain assertion error, but simply propagate the RuntimeError up...

Will discard debug commits shortly.

**Comment by leofang** [comment]
> Will discard debug commits shortly.

Done. 

Jenkins, test this please

File: tests/cupy_tests/test_cudnn.py [hunk 1]
```diff
@@ -333,6 +334,11 @@ def setUp(self):
         if ((self.dilate > 1 and version < 6000) or
                 (self.groups > 1 and version < 7000)):
             self.err = ValueError
+        elif (sys.platform.startswith('win32') and version == 7605
+                and deterministic and self.dtype == numpy.float16
+                and self.ndim == 3 and self.dilate == 2 and self.groups == 2):
```
Comment (leofang) [inline tests/cupy_tests/test_cudnn.py:339]: This is a very stringent skip condition, just enough to skip the failing tests in the Windows CI. I checked cuDNN's documentation and release notes, but I don't think a Windows-only failure like this is expected, so I'd prefer to make the skip range narrow. If we expand the Windows CI to test more versions, we will have a better clue and can revisit this.

**Comment by leofang** [comment]
@emcastillo The Windows CI is all green now 🎉

**Comment by leofang** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Pull request merged**